dspy/utils/asyncify.py

```diff
@@ -0,0 +1,33 @@
+import sys
+from typing import Awaitable, Optional, TypeVar, Callable
+
+from anyio import CapacityLimiter
+from asyncer import asyncify as __asyncify
+
+if sys.version_info >= (3, 10):
+    from typing import ParamSpec
+else:
+    from typing_extensions import ParamSpec
+
+T_Retval = TypeVar("T_Retval")
+T_ParamSpec = ParamSpec("T_ParamSpec")
+T = TypeVar("T")
+
+
+class AsyncLimiter:
+    _limiter: Optional[CapacityLimiter] = None
+
+    @classmethod
+    def get(cls):
+        import dspy
+
+        if cls._limiter is None or cls._limiter.total_tokens != dspy.settings.async_capacity:
+            cls._limiter = CapacityLimiter(dspy.settings.async_capacity)
+
+        return cls._limiter
+
+
+def asyncify(
+    program: Callable[T_ParamSpec, T_Retval],
+) -> Callable[T_ParamSpec, Awaitable[T_Retval]]:
```

Comment:
The type annotation part is not really readable. I am never a fan of Python type annotation, so I would rather just delete the type annotation for simplicity. But will let @okhat make the decision.

Reply:
Removed!